test/functionality/dataflow/elements/loops/while-loop.ts

```diff
@@ -0,0 +1,55 @@
+import { assertDataflow, withShell } from '../../../helper/shell'
+import { DataflowGraph, EdgeType, initializeCleanEnvironments } from '../../../../../src/dataflow'
+import { define } from '../../../../../src/dataflow/environments'
+import { LocalScope } from '../../../../../src/dataflow/environments/scopes'
+
+describe('while', withShell(shell => {
+	assertDataflow('simple constant while', shell,
+		'while (TRUE) 2',
+		new DataflowGraph()
+	)
+	assertDataflow('using variable in body', shell,
+		'while (TRUE) x',
+		new DataflowGraph().addVertex({ tag: 'use', id: '1', name: 'x', when: 'maybe' })
+	)
+	assertDataflow('assignment in loop body', shell,
+		'while (TRUE) { x <- 3 }',
+		new DataflowGraph().addVertex({ tag: 'variable-definition', id: '1', name: 'x', scope: LocalScope, when: 'maybe' })
+	)
+	assertDataflow('def compare in loop', shell, 'while ((x <- x - 1) > 0) { x }',
+		new DataflowGraph()
+			.addVertex({ tag: 'variable-definition', id: '0', name: 'x', scope: LocalScope })
+			.addVertex({ tag: 'use', id: '1', name: 'x' })
+			.addVertex({ tag: 'use', id: '7', name: 'x', when: 'maybe', environment: define({ name: 'x', nodeId: '0', definedAt: '4', used: 'always', kind: 'variable', scope: LocalScope }, LocalScope, initializeCleanEnvironments()) })
+			.addEdge('7', '0', EdgeType.Reads, 'maybe')
+			.addEdge('0', '1', EdgeType.DefinedBy, 'always')
+	)
+	assertDataflow('Endless while loop',
+		shell,
+		'while(TRUE) 1',
+		new DataflowGraph()
+	)
+	assertDataflow('Endless while loop with variables',
+		shell,
+		'while(x) y',
+		new DataflowGraph()
+			.addVertex({ tag: 'use', id: '0', name: 'x', when: 'always' })
+			.addVertex({ tag: 'use', id: '1', name: 'y', when: 'maybe' })
+	)
+	assertDataflow('while loop with updated variable',
+		shell,
+		'x <- 1 ; while(x < 2) {x <- x + 1}',
+		new DataflowGraph()
+			.addVertex({ tag: 'variable-definition', id: '0', name: 'x', when: 'always', scope: LocalScope})
+			.addVertex({ tag: 'variable-definition', id: '6', name: 'x', when: 'maybe', scope: LocalScope})
```

Comment:
Investigating the logging output reveals:

```

2023-11-10 14:21:21.614	WARN	/src/dataflow/environments/environment.ts:151	main:dataflow	Comparing undefined environments undefined and undefined
2023-11-10 14:21:21.615	WARN	/src/dataflow/environments/environment.ts:155	main:dataflow	Different environments: .GlobalEnv!=.GlobalEnv or 3!=4; {"name":".GlobalEnv","id":"0","memory":{}} and {"name":".GlobalEnv","id":"51","memory":{}} due to different names or sizes ([["return",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"return","nodeId":"built-in"}]],["cat",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"cat","nodeId":"built-in"}]],["print",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"print","nodeId":"built-in"}]]] vs. [["return",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"return","nodeId":"built-in"}]],["cat",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"cat","nodeId":"built-in"}]],["print",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"print","nodeId":"built-in"}]],["x",[{"nodeId":"0","scope":"local","name":"x","used":"always","kind":"variable","definedAt":"2"}]]])
2023-11-10 14:21:21.615	WARN	/src/dataflow/environments/environment.ts:183	main:dataflow	Different environments {"current":{"name":".GlobalEnv","id":"0","memory":{}},"level":0} and {"current":{"name":".GlobalEnv","id":"51","memory":{}},"level":0}
2023-11-10 14:21:21.615	WARN	/src/dataflow/graph/equal.ts:83	main:dataflow	node 6 does not match on environments ({"current":{"name":".GlobalEnv","id":"0","memory":{}},"level":0} vs {"current":{"name":".GlobalEnv","id":"51","memory":{}},"level":0})
```

Which shows that on vertex 6 it exepcted `["x",[{"nodeId":"0","scope":"local","name":"x","used":"always","kind":"variable","definedAt":"2"}]]` to be part of the definition as well. In other words: by your initial definition of `x <- 1` you define an environment that can be seen by the "other `x`es as well. 

Does investigating the log help you in understanding the presence/absence of the environment definitions? Or how would you want them to behave?

Reply:
So with the new differ of #493, you would get the following result as part of the guard failure error message. is this more helpful?

```
 * Vertex 6 differs in environments. Different environment sizes. expected: {"name":".GlobalEnv","id":"0","memory":[["return",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"return","nodeId":"built-in"}]],["cat",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"cat","nodeId":"built-in"}]],["print",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"print","nodeId":"built-in"}]]]} vs. got: {"name":".GlobalEnv","id":"37","memory":[["return",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"return","nodeId":"built-in"}]],["cat",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"cat","nodeId":"built-in"}]],["print",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"print","nodeId":"built-in"}]],["x",[{"nodeId":"0","scope":"local","name":"x","used":"always","kind":"variable","definedAt":"2"}]]]}
 * Vertex 6 differs in environments. Key comparison. More in got: ["x"]
 * Vertex 3 differs in environments. Different environment sizes. expected: {"name":".GlobalEnv","id":"0","memory":[["return",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"return","nodeId":"built-in"}]],["cat",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"cat","nodeId":"built-in"}]],["print",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"print","nodeId":"built-in"}]]]} vs. got: {"name":".GlobalEnv","id":"35","memory":[["return",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"return","nodeId":"built-in"}]],["cat",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"cat","nodeId":"built-in"}]],["print",[{"kind":"built-in-function","scope":".GlobalEnv","used":"always","definedAt":"built-in","name":"print","nodeId":"built-in"}]],["x",[{"nodeId":"0","scope":"local","name":"x","used":"always","kind":"variable","definedAt":"2"}]]]}
 * Vertex 3 differs in environments. Key comparison. More in got: ["x"]
 * Vertex 7 differs in environments. Different ids for x. expected: 6 vs. got: 0
 * Vertex 7 differs in environments. Different used for x. expected: maybe vs. got: always
 * Vertex 7 differs in environments. Different definition ids (definedAt) for x. expected: 0 vs. got: 2
 ```